qiskit/circuit/exceptions.py

```diff
@@ -19,3 +19,9 @@ class CircuitError(QiskitError):
     """Base class for errors raised while processing a circuit."""
 
     pass
+
+
+class ParameterTypeError(TypeError):
+    """The type error throw by parameter if we try to cast to number unbind parameter."""
+
+    pass
```

Comment:
I didn't express myself clearly. I know that I can have multiple inheritances, and it works correctly in the general case. But not in our case. The issue lay in the operator.py and in catching the QiskitError there. If ParameterTypeError inherits from Qiskit error it is caught there and not in template optimization. And in a result, the tests fail.

![image](https://user-images.githubusercontent.com/55279376/152678821-6fa0cf58-ddf9-44fb-ad10-0f6afb0c66d2.png)

Reply:
So If we would like to inherit from QiskitError I have to change the operator.py and catch fewer general exceptions or change in some way our code in template optimization. Therefore I prefer to inherit only from type errors. @jakelishman  Do you agree?

Reply:
Ah, I understand your issue - that's pretty awkward programming in `operator.py` to be honest, and I wouldn't really know what to suggest.  Fortunately, I think we can sidestep a lot of this, because the changes to `UnitaryGate` became unnecessary, so we can just remove the exception as well.